### ISSUE TITLE
Exception running MergeBamAlignment after #622

### ISSUE THREAD

**Issue by tfenne** [broadinstitute/picard#634]
Hey @nh13, @bradtaylor , @yfarjoun I'm running into issues running MergeBamAlignment after #622 was merged in.  What appears to be happening is that the code to re-calculate the NM and MD tags is running on reads that are query-sorted, but using a ReferenceSequenceFileWalker to get access to the reference bases, so as soon as it encounters a read on an earlier reference it blows up with something like this:

```
Exception in thread "main" htsjdk.samtools.SAMException: Requesting earlier reference sequence: 0 < 8
       	at htsjdk.samtools.reference.ReferenceSequenceFileWalker.get(ReferenceSequenceFileWalker.java:82)
       	at picard.sam.AbstractAlignmentMerger.calculateMdAndNmTags(AbstractAlignmentMerger.java:596)
       	at picard.sam.AbstractAlignmentMerger.clipForOverlappingReads(AbstractAlignmentMerger.java:628)
       	at picard.sam.AbstractAlignmentMerger.transferAlignmentInfoToPairedRead(AbstractAlignmentMerger.java:580)
       	at picard.sam.AbstractAlignmentMerger.mergeAlignment(AbstractAlignmentMerger.java:390)
       	at picard.sam.SamAlignmentMerger.mergeAlignment(SamAlignmentMerger.java:152)
       	at picard.sam.MergeBamAlignment.doWork(MergeBamAlignment.java:266)
       	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:208)
       	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:95)
       	at picard.cmdline.PicardCommandLine.main(PicardCommandLine.java:105)
```

**Comment by yfarjoun** [comment]
Oh man! I should've thought of this. This is exactly why we have
SetNmAndUqTags, I guess we should also update MD in this same program.

Could you revert that commit please?

**Comment by nh13** [comment]
I will make another commit to add the calculation of `fixNMandUQ` instead

**Comment by nh13** [comment]
@yfarjoun @tfenne 
I see a few choices:
1. Revert it: https://github.com/broadinstitute/picard/pull/635
2. Fix it: https://github.com/broadinstitute/picard/pull/636/
3. Revert it, then fix it.

Your call, my apologies for introducing the bug!

**Comment by yfarjoun** [comment]
if you can fix it soon, I'll go with #2...but it should include a small
test case of queryname sorted bam to be sure... otherwise, I think it would
be best to revert so that you have time to submit a non-buggy version.

Y.

**Comment by nh13** [comment]
@yfarjoun I added running `SamAlignmentMerger` with `queryname` sort to an existing test.  I am surprised there wasn't one there already.
